src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,394 @@
+
+                    }
+                    else if (e is RemoteException { Error: RemoteError.DistributedObjectDestroyed })
+                    {
+                        throw new DistributedObjectDestroyedException(e);
+                    }
+                    else if (e is RemoteException { Error: RemoteError.WaitKeyCancelledException })
+                    {
+                        _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+                        return InvalidFence;
+                    }
+                }
+                catch
+                {
+                    _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+                    throw;
+                }
+            }
+        }
+
+        /// <inheritdoc/> 
+        public async Task<bool> TryLockAsync(TimeSpan timeout)
+        {
+            var fence = await TryLockAndGetFenceAsync(timeout).CfAwait();
+            return fence != InvalidFence;
+        }
+
+        /// <inheritdoc/> 
+        public async Task<bool> TryLockAsync()
+        {
+            var fence = await TryLockAndGetFenceAsync(TimeSpan.FromMilliseconds(0)).CfAwait();
+            return fence != InvalidFence;
+        }
+
+        /// <inheritdoc/> 
+        public async Task UnlockAsync()
+        {
+            var threadId = ContextId;
+            var sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            if (sessionId == CPSubsystemSession.NoSessionId)
+            {
+                _lockedThreadToSession.TryRemove(threadId, out var _);
+                throw new SynchronizationLockException();
+            }
+
+            try
+            {
+                Guid invocationId = Guid.NewGuid();
+                bool stillLockedByCurrentThread = await RequestUnlockAsync(sessionId, threadId, invocationId).CfAwait();
+
+                if (stillLockedByCurrentThread)
+                    _lockedThreadToSession[threadId] = sessionId;
+                else
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+
+                _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+            }
+            catch (RemoteException e)
+            {
+                if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                {
+                    _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+                    throw new SessionExpiredException(e);
+                }
+                else if (e is RemoteException { Error: RemoteError.IllegalMonitorState })
+                {
+                    _lockedThreadToSession.TryRemove(threadId, out var _);
+                    throw new SynchronizationLockException("IllegalMonitorState exception is thrown by server. Are you trying to unlock without holding?", e);
+                }
+            }
```

Comment:
should you throw for all other `RemoteException` cases?

Reply:
Java client throws only for these cases. I have followed that.

Reply:
Still will be changed as Java.

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/993cabbcb9bac708cc38ea83239b98cf19ead875